### PR TITLE
[alts]: store tsi peer property in dynamic metadata.

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
Signed-off-by: chaoqin-li1123 <chaoqinli@google.com>

Commit Message: get the peer property from alts handshake validator and set the dynamic metadata.
Additional Description: NA
Risk Level: low
Testing: NA
Docs Changes: NA
Release Notes: NA
Platform Specific Features:NA
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/transport_sockets/alts/tsi_socket.cc [hunk 1]
```diff
@@ -106,14 +106,20 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
...
       } else {
         ENVOY_CONN_LOG(debug, "TSI: Handshake validation failed: {}", callbacks_->connection(),
                        err);
         return Network::PostIoAction::Close;
       }
+      ProtobufWkt::Struct dynamic_metadata;
+      ProtobufWkt::Value val;
+      val.set_string_value(tsi_info.name_);
+      dynamic_metadata.mutable_fields()->insert({std::string("peer_name"), val});
+      callbacks_->connection().streamInfo().setDynamicMetadata("envoy.alts.tsi", dynamic_metadata);
```
Comment (lizan) [inline source/extensions/transport_sockets/alts/tsi_socket.cc:122]: document this dynamic metadata in ALTS docs?

**Review by lizan — COMMENTED** [review]
can you add a release note as well?

File: source/extensions/transport_sockets/alts/tsi_socket.cc (hunk 1 above)
Comment (chaoqin-li1123) [reply source/extensions/transport_sockets/alts/tsi_socket.cc:122]: alts don't have its own rst file, only a proto. Am I supposed to document it its proto file?

File: source/extensions/transport_sockets/alts/config.cc [hunk 2]
```diff
@@ -38,7 +38,8 @@ void grpcAltsSetRpcProtocolVersions(grpc_gcp_rpc_protocol_versions* rpc_versions
 // Returns true if the peer's service account is found in peers, otherwise
 // returns false and fills out err with an error message.
 bool doValidate(const tsi_peer& peer, const absl::node_hash_set<std::string>& peers,
-                std::string& err) {
+                TsiInfo& tsi_info, std::string& err) {
+  tsi_info.name_ = TSI_ALTS_SERVICE_ACCOUNT_PEER_PROPERTY;
```
Comment (yanavlasov) [inline source/extensions/transport_sockets/alts/config.cc:42]: You should set it to the `value` variable down inside the the `if (name.compare(TSI_ALTS_SERVICE_ACCOUNT_PEER_PROPERTY) ... ` statement.

**Comment by yanavlasov** [comment]
I think we should also add verification that dynamic metadata was populated in the [alts integration test](https://github.com/envoyproxy/envoy/blob/5a2d484382b36251aaa131179a056c1bce850a7f/test/extensions/transport_sockets/alts/alts_integration_test.cc#L248)

You can add configuration to the router to add a header containing a value from dynamic metadata in the [route config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto#envoy-v3-api-field-config-route-v3-routeconfiguration-request-headers-to-add) and verify that the server gets this header.

/wait

**Review by yanavlasov — CHANGES_REQUESTED** [review]

File: source/extensions/transport_sockets/alts/tsi_socket.cc (hunk 1 above)
Comment (yanavlasov) [reply source/extensions/transport_sockets/alts/tsi_socket.cc:122]: Yes, I think adding a line in a proto file should be fine.

File: source/extensions/transport_sockets/alts/tsi_socket.cc [hunk 3]
```diff
@@ -106,14 +106,20 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
     }
     if (handshake_validator_) {
       std::string err;
-      const bool peer_validated = handshake_validator_(peer, err);
+      TsiInfo tsi_info;
+      const bool peer_validated = handshake_validator_(peer, tsi_info, err);
```
Comment (antoniovicente) [inline source/extensions/transport_sockets/alts/tsi_socket.cc:110]: Would it be possible to add tests to validate the functional changes introduced by this PR?

Comment (chaoqin-li1123) [reply source/extensions/transport_sockets/alts/tsi_socket.cc:110]: yes, I am still trying to figure out how to verify the dynamic metadata is actually set.

**Comment by chaoqin-li1123** [comment]
Because this is in transport socket, the dynamic metadata can't be verified by access log(only works for http). I tried to send the metadata in upstream request, but it somehow doesn't work. The dynamic metadata can't be added to metadata. Do you have any idea why it doesn't work? I check the test log, setting the dynamic metadata is before the request encoding complete, so it should be encoded.

**Comment by chaoqin-li1123** [comment]
callbacks_->streamInfo() is actually different from callbacks_->connection().streamInfo()... so the dynamic metadata set in callbacks_->connection().streamInfo() doesn't get into request header in 
  route_entry_->finalizeRequestHeaders(headers, callbacks_->streamInfo(),
                                       !config_.suppress_envoy_headers_);

File: api/envoy/extensions/transport_sockets/alts/v3/alts.proto [hunk 4]
```diff
@@ -15,6 +15,7 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#extension: envoy.transport_sockets.alts]
 
 // Configuration for ALTS transport socket. This provides Google's ALTS protocol to Envoy.
+// Store the tsi peer name in dynamic metadata.
```
Comment (yanavlasov) [inline api/envoy/extensions/transport_sockets/alts/v3/alts.proto:18]: I would also add the key name in the metadata and also document that you are storing Struct under that key with the "peer_identity" field, so people do not need to go hunting for it in the source code.

File: api/envoy/config/transport_socket/alts/v2alpha/alts.proto [hunk 5]
```diff
@@ -17,6 +17,7 @@ option (udpa.annotations.file_status).package_version_status = FROZEN;
 // [#extension: envoy.transport_sockets.alts]
 
 // Configuration for ALTS transport socket. This provides Google's ALTS protocol to Envoy.
+// Store the tsi peer name in dynamic metadata.
```
Comment (yanavlasov) [inline api/envoy/config/transport_socket/alts/v2alpha/alts.proto:20]: I do not think we still need to edit the v2alpha protos.

File: source/extensions/transport_sockets/alts/tsi_socket.cc [hunk 6]
```diff
@@ -106,14 +106,22 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
...
       } else {
         ENVOY_CONN_LOG(debug, "TSI: Handshake validation failed: {}", callbacks_->connection(),
                        err);
         return Network::PostIoAction::Close;
       }
+      ProtobufWkt::Struct dynamic_metadata;
+      ProtobufWkt::Value val;
+      val.set_string_value(tsi_info.name_);
+      dynamic_metadata.mutable_fields()->insert({std::string("peer_identity"), val});
+      callbacks_->connection().streamInfo().setDynamicMetadata("envoy.transport_sockets.alts",
```
Comment (yanavlasov) [inline source/extensions/transport_sockets/alts/tsi_socket.cc:122]: I think the metadata key can be more generic, such as `envoy.transport_socket.peer_information`. For instance peer identity can also be set by TLS transport with mTLS support.

**Review by yanavlasov — CHANGES_REQUESTED** [review]

**Comment by yanavlasov** [comment]
I think the change looks very good. I would just improve the doc a little and also make the metadata key more generic since this metadata also applies to other transport sockets.

/wait

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**